airflow/api_fastapi/execution_api/routes/task_instance.py

```diff
@@ -0,0 +1,152 @@
+    passed along. (Check our the schemas for details, the rendered docs might not reflect this accurately)
+    """
+    # We only use UUID above for validation purposes
+    ti_id_str = str(task_instance_id)
+
+    old = select(TI.state).where(TI.id == ti_id_str).with_for_update()
+    try:
+        (previous_state,) = session.execute(old).one()
+    except NoResultFound:
+        log.error("Task Instance %s not found", ti_id_str)
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail={
+                "reason": "not_found",
+                "message": "Task Instance not found",
+            },
+        )
+
+    # We exclude_unset to avoid updating fields that are not set in the payload
+    data = ti_patch_payload.model_dump(exclude_unset=True)
+
+    query = update(TI).where(TI.id == ti_id_str).values(data)
+
+    if isinstance(ti_patch_payload, schemas.TIEnterRunningPayload):
+        if previous_state != State.QUEUED:
+            log.warning(
+                "Can not start Task Instance ('%s') in invalid state: %s",
+                ti_id_str,
+                previous_state,
+            )
+
+            # TODO: Pass a RFC 9457 compliant error message in "detail" field
+            # https://datatracker.ietf.org/doc/html/rfc9457
+            # to provide more information about the error
+            # FastAPI will automatically convert this to a JSON response
+            # This might be added in FastAPI in https://github.com/fastapi/fastapi/issues/10370
+            raise HTTPException(
+                status_code=status.HTTP_409_CONFLICT,
+                detail={
+                    "reason": "invalid_state",
+                    "message": "TI was not in a state where it could be marked as running",
+                    "previous_state": previous_state,
+                },
+            )
+        log.info("Task with %s state started on %s ", previous_state, ti_patch_payload.hostname)
+        # Ensure there is no end date set.
+        query = query.values(
+            end_date=None,
+            hostname=ti_patch_payload.hostname,
+            unixname=ti_patch_payload.unixname,
+            pid=ti_patch_payload.pid,
+            state=State.RUNNING,
+        )
+    elif isinstance(ti_patch_payload, schemas.TITerminalStatePayload):
+        # If state is skipped, we need to set the end_date to the start_date if it is not set
+        end_date = ti_patch_payload.end_date
+        if ti_patch_payload.state == State.SKIPPED and not ti_patch_payload.end_date:
+            query = query.values(end_date=TI.start_date, duration=0)
+        elif ti_patch_payload.state in [State.UPSTREAM_FAILED, State.REMOVED]:
+            # We do not update any other fields for these states
+            # as we do not expect it to even reach the worker.
+            # The scheduler should handle these states.
+            pass
+        else:
+            if not end_date:
```

Comment:
Minor nit: Make this `elif not end_date:`?

Reply:
Should this be handled as a model validation on the schema perhaps? That way all validation logic remains on the schema keeping the controller simpler?

Reply:
Simplified it and removed this in https://github.com/apache/airflow/pull/43602/commits/fa4f2228a2a057d7252e2d1dda64e3575f4b8b94